create-react-app で メモアプリ(SPA版)を作成

GennyBoy opened the pull request:
[ReactでSPAを作る](https://bootcamp.fjord.jp/practices/262)のプラクティスの課題です。
ご確認よろしくお願いいたします。

GennyBoy committed:
localStorage に保存されているデータを一覧に表示できるようにした

やったこと
- List コンポーネントの作成

やってないこと
- 細かいスタイルの調整（リストの先頭の・の非表示など）

GennyBoy committed:
メモ追加機能を実装

やったこと
- +ボタンを押したら入力モードに移行する
- テキストを入力して、「編集」ボタンを押したら、該当メモが一覧に追加される

やってないこと
- メモ追加後に一覧モードに移行する処理の追加
- 「削除」ボタンを押下した際の処理の追加
- レイアウトの調整

GennyBoy committed:
重複している処理を共通化した

GennyBoy committed:
データ追加する機能を実装して該当コードが不要になったので削除

GennyBoy committed:
編集ボタンをクリックした後に一覧表示モードに移行するようにした

GennyBoy committed:
メモの本文を何も入力せずに編集ボタンをクリックしたらアラートを表示するようにした

GennyBoy committed:
メモの入力欄を開いて何もせずに空のまま編集ボタンを押すと前に入力したメモの内容で保存されてしまうバグを修正

入力欄を開いてから、何もせずにsubmitすると、入力欄が空にも関わらず `memo` に一個前に保存した文章が残っているため一個前に入力した文章でメモが追加されてしまっていた。
何らかの入力をした後に、削除して空にしてからsubmitすると `onChange` が走るので問題なくアラートが出る。
編集ボタンを押して、メモの一覧を update した後に、 memo の state を空文字で更新することでこの問題を解消した。

GennyBoy committed:
メモの編集機能を追加

- `memo` state の値をオブジェクトに変更して `id` も持たせるようにした
  - 編集時のデータ更新に必要
- `isAdding` state を導入
  - `handleEditButtonClick` 内の処理を編集と新規追加で分けた

GennyBoy committed:
メモ編集モード時に + を押下すると、新規追加モードに移行するようにした

GennyBoy committed:
編集しているメモは一覧で非リンクとして表示するようにした

GennyBoy committed:
削除機能を実装

GennyBoy committed:
メモ一覧の見た目をモックに近付けた

GennyBoy committed:
class名を一般的な命名規則に沿ったものにした

GennyBoy committed:
見た目をモックに近づけた

GennyBoy committed:
JSX内の複雑な処理をhandlerとして抽出

GennyBoy committed:
state 間で不整合な状態にならないようにするため不必要なstateを削除

GennyBoy committed:
npm install --save-dev --save-exact prettier

GennyBoy committed:
npx prettier . --write

GennyBoy committed:
npx eslint --init

GennyBoy committed:
今回はテストは考慮しないため削除

GennyBoy committed:
npx eslint . --fix

GennyBoy committed:
eslint のエラーを解消

GennyBoy committed:
エラーの解消

https://stackoverflow.com/questions/76435306/babel-preset-react-app-is-importing-the-babel-plugin-proposal-private-propert

GennyBoy committed:
eslint 'no-alert' エラーの対応

`alert` を使うと ESLint に怒られるので ( https://eslint.org/docs/latest/rules/no-alert )、
`alert` を使わずにエラーメッセージを表示するようにした

GennyBoy committed:
status 名が状態を正しく表現するように変更

GennyBoy committed:
memo status をクリアする処理を共通化

GennyBoy committed:
重複しているコードを削除

GennyBoy committed:
コンポーネント分割しやすいようにhandlerを関数として抽出した

GennyBoy committed:
今回のプラクティスでは標準のルールだけ適用する

GennyBoy committed:
prop-types のルールを無効にした

実際には TypeScriptを使うことが多いと思うので、今回はこのルールはOFFにした

GennyBoy committed:
メモ編集エリアを react コンポーネントに分割

GennyBoy committed:
npx prettier . --write

GennyBoy committed:
EditMemoPanel コンポーネントを別ファイルに分割

GennyBoy committed:
関数名を自明にした

GennyBoy committed:
コメント追加

GennyBoy committed:
要件にある仕様に合わせた

GennyBoy committed:
'+' ボタンを押下すると同時に新規メモがDBにも保存されることにした

この対応を入れる前は'+'を押すと左ペインのリストには、新規メモが表示されるが
DBには保存されていないのでリロードすると消えてしまう現象が起きていた。

GennyBoy committed:
メモ本文の一文目をタイトルとして抽出する処理を共通化した

GennyBoy committed:
データ不整合を起こさないために、DBに保存する処理を先に実行することにした

GennyBoy committed:
コンポーネントの親子関係を実態に沿う形に変更

App 配下に List と EditMemoPanel を siblings として配置した
それに伴い全ての state をリフトアップした

GennyBoy committed:
メモのタイトルを本文から抽出する処理を共通化

GennyBoy committed:
useEffect を使う練習

GennyBoy committed:
adding という status が必要なくなったため state の名前を変更した

これまでは削除ボタンを新規メモ追加時には非表示にしていた。
なぜなら、新規メモフォーム表示時には、該当メモはまだStorageには保存しておらず、
編集ボタンをクリック時にStorageに該当メモを保存する仕様にしていたため、新規メモ追加フォーム
において、削除というアクションは存在しなかったから。
なのでアプリの status として、 adding/editing/viewing の３つを区別する必要があった。

しかし仕様を変更して、新規メモフォーム表示時（正確には `+` ボタンをクリックしたタイミングで）
には該当メモをstorageに保存するようにした。
こうすると新規メモフォーム表示時にも削除ボタンが表示されていることは何ら不自然ではないので
そうした。
上記対応を取ると、 元々存在した `status` という state を `isEditing` とした方が自明な上、
boolean で表現できて誤用の可能性を減らせるためそうした。

GennyBoy committed:
README.md の中身を書いた

GennyBoy committed:
prettier 実行時に README.md を ignore する

GennyBoy committed:
npx prettier . --write

GennyBoy committed:
ファイルの末尾に改行を追加

GennyBoy committed:
不要な dependency を削除

summertree128 commented on src/EditMemoPanel.js:
```diff
@@ -0,0 +1,72 @@
+    const id = memo.id;
+    const { content } = memo;
+    const title = extractTitleFromContent(content);
+    localStorage.setItem(id, content);
```
Local Storageというのはさまざまな用途で使われるものですので、このようなデータ設計はあまりよくないかなと思いました。memo以外のものも格納されるという前提に立ったほうが良いです。
まずkeyを見てそこに何のデータが入っているのか分かるようなkey nameにしましょう。

GennyBoy replied:
> memo以外のものも格納されるという前提に立ったほうが良いです。

こちら全く意識できていませんでした 💦 
下記のように、 Prefix として `memo-` をつけるように変更しました。
https://github.com/GennyBoy/react-memo-spa/blob/f1c21d52cca1f55c9671e9cac1b8f65b4e9e1184/src/List.js#L12

summertree128 commented on src/App.js:
```diff
@@ -1,9 +1,54 @@
+  const [isEditing, setIsEditing] = useState(false);
+  const [memos, setMemos] = useState(() => {
+    const memoList = [];
+    for (let i = 0; i < localStorage.length; i += 1) {
```
localStorageにメモしか入っていないことが前提になっているようですが、localStorageというのは色々な用途で使われるものです。メモ以外のデータも格納されることがあるという前提に立ったほうが良いです。

まずkeyを見てそこに何のデータが入っているのか分かるようなkey nameにしましょう。
その上で必要なkeyに対してのみ処理を行うべきです。

summertree128 replied:
すみません、https://github.com/GennyBoy/react-memo-spa/pull/1#discussion_r1657557827 とほぼ同じようなコメントを書いてしまいました 😓 
どちらか片方は消したつもりだったのですがうまく消せていなかったようです。
言いたいことは同じなので、localStorageの再設計をお願いしますー

GennyBoy replied:
こちらもご指摘ありがとうございます！

下記のように、Prefix に `memo-` とついているものだけに対して処理するようにいたしました。
https://github.com/GennyBoy/react-memo-spa/blob/f1c21d52cca1f55c9671e9cac1b8f65b4e9e1184/src/App.js#L16-L25

summertree128 replied:
これだと根本的には変わっていないというか、LocalStorageに格納されている全てのkeyを順番にチェックしていくという処理のままですね。memoの他にもたくさんkeyがあったりすると効率が良くないので、必要なkeyだけを取得するような処理（とデータ設計）にしてください。

また、`localStorage.key()`を使って上から順番に取得していく処理だと、メモの順番がブラウザの実装依存になってしまうのであまり良くない気がします。（順番を気にしないのならいいですが、流石にそういうわけにはいかないかなと）
参考： https://developer.mozilla.org/en-US/docs/Web/API/Storage/key#:~:text=The%20order%20of%20keys%20is%20user%2Dagent%20defined%2C%20so%20you%20should%20not%20rely%20on%20it.

summertree128 replied:
というか、今やってみたら、メモが複数ある状態で新規メモを登録してからリロードすると順番が変わる時がありますね。ちょっと不自然な動きだと感じました。
(おそらく、Chromeはキーを辞書順にソートしているからだと思います）

summertree128 replied:
あと、これは本質的なところではないですが、普通はこう書くと思います。
```suggestion
    for (let i = 0; i < localStorage.length; i++) {
```

summertree128 commented on src/App.js:
```diff
@@ -1,9 +1,54 @@
+
+  // useEffect を使う練習用のコード
+  useEffect(() => {
+    console.log(`LocalStorageに保存されているメモの数: ${localStorage.length}`);
```
`useEffect`の練習ができると良いというのはその通りなのですが、これはちょっと意味がなさすぎるというか、普通はこういう出力はしませんね。例えば`useEffect`の中でlocalStorageからメモを読み込むとかそういうのをイメージしていました。何かしら意味のある処理をしてください。

GennyBoy replied:
下記のように、 localStorage からメモを読み込む処理を useEffect のなかに記述してみました。
https://github.com/GennyBoy/react-memo-spa/blob/f1c21d52cca1f55c9671e9cac1b8f65b4e9e1184/src/App.js#L12-L28

summertree128 commented on src/App.js:
```diff
@@ -1,9 +1,54 @@
+        memo={memo}
+        updateMemo={setMemo}
+      />
+      {isEditing === true && (
```
`=== true`はなくても大丈夫です！truthyならそれで十分です〜

GennyBoy replied:
修正しました。
86ff17fe8b5cfec5cf88145558311bf8fa993b83

summertree128 commented on src/EditMemoPanel.js:
```diff
@@ -0,0 +1,72 @@
+  memo,
+  updateMemo,
+}) {
+  function clearMemoState() {
```
stateというといわゆるReactのstateを想像するので、なるべく避けたほうがいい単語かな〜と思いました。

GennyBoy replied:
`clearMemoInEdit` に変更しました
https://github.com/GennyBoy/react-memo-spa/blob/f1c21d52cca1f55c9671e9cac1b8f65b4e9e1184/src/EditMemoPanel.js#L8-L10

summertree128 commented on src/EditMemoPanel.js:
```diff
@@ -0,0 +1,72 @@
+        type="submit"
+        onClick={handleEditButtonClick}
+      >
+        編集
```
個人的な意見ですが、これは編集というより「保存」とかそんな感じかなと思いました。（編集だと、今から編集画面が立ち上がるボタンかなと想像しました）

GennyBoy replied:
確かに「保存」の方がボタンが押した後に起こることを正確に表現できそうです 💡
ありがとうございます！
26d57be33a88032599d12b3438da7a018f2f8861

summertree128 commented on src/App.js:
```diff
@@ -1,9 +1,54 @@
+        <EditMemoPanel
+          isEditing={isEditing}
+          updateIsEditing={setIsEditing}
+          memos={memos}
+          updateMemos={setMemos}
+          memo={memo}
+          updateMemo={setMemo}
```
`memo`と`memos`が紛らわしいような気がしますね。この名前だと、そもそも`memo`は必要なさそうに感じます。
実際には`memo`はただのメモというより現在編集中のメモとかそんな感じだと思うので、そういう名前にしたほうがいと思います。

GennyBoy replied:
ご指摘ありがとうございます！
下記のように `memoInEdit` に変更しました。
https://github.com/GennyBoy/react-memo-spa/blob/f1c21d52cca1f55c9671e9cac1b8f65b4e9e1184/src/App.js#L8-L9

summertree128 commented on src/List.js:
```diff
@@ -0,0 +1,54 @@
+    const id = window.self.crypto.randomUUID();
+    const content = "新規メモ";
+    const title = extractTitleFromContent(content);
+    localStorage.setItem(id, content);
+    updateMemo({ id: id, content: content });
+    updateMemos([
```
`localStorage.setItem`と`updateMemos`はどちらもメモの保存（ローカルストレージに保存するかオンメモリのstateに保存するかという違うはあれど）だと思うので、ひとまとめの処理になっていたほうがうれしかなと思いました。

この`List`コンポーネントの視点に立つと、保存先のストレージがどこかというのは気にしなくていい詳細かなと思います。

GennyBoy replied:
state と ストレージへの保存(または削除)を関数にまとめて、コンポーネントからはこれらの関数を呼び出すように修正しました
https://github.com/GennyBoy/react-memo-spa/blob/f1c21d52cca1f55c9671e9cac1b8f65b4e9e1184/src/App.js#L30-L52

summertree128 reviewed (commented):
レビューしました！とてもよくできていると思います〜。

もう少し良くできそうな部分があると思ったのでいくつかコメントしています。ご確認ください。

GennyBoy committed:
localStorage には今回のApp以外のデータがあるという前提でコードを修正

- localStorage にデータを保存するとき、 `memo-` という Prefix をつけることにした
- localStorage に保存されているデータを、メモ一覧に表示するときに key に該当の Prefix がついていない場合は処理の対象から外した

GennyBoy committed:
練習用に書いた useEffect のなかでもっと意味のある処理をするようにした

GennyBoy committed:
false でければ(= truthy な値であれば)OKなので、そのような条件文にした

GennyBoy committed:
Reactのstateとの混同を避けるため関数名を変更

GennyBoy committed:
使われていない props を削除

GennyBoy committed:
ボタンのラベルをより適切な表現に変更

GennyBoy committed:
memo と memos が紛らわしいので、 memo の変数名を現在編集中のメモという意味がわかるように変更

GennyBoy committed:
編集中のメモを保存した時の処理を関数にして、関数を使う側ではデータの保存先（ state とストレージ）を意識しないでいいようにした

GennyBoy committed:
npx prettier . --write

summertree128 commented on src/EditMemoPanel.js:
```diff
@@ -0,0 +1,56 @@
+      >
+        削除
+      </button>
+      {memoInEdit.content === "" ? (
+        <div className="error-message">空のメモは保存できません</div>
+      ) : null}
```
ある条件を満たすときだけ要素を表示する、というのは下記のように書くのが通例です。
```suggestion
      {memoInEdit.content === "" && <div className="error-message">空のメモは保存できません</div>}
```

GennyBoy replied:
ありがとうございます。ご指摘いただいた点修正してみました 🙇‍♂️ 

**localStorage のデータ設計**
987259a24aebad318907228cecd87d5151a3d819

**インクリメントの書き方を修正**
81482e9c2ce103158545bc8102b47222dde82a0d

**条件によって表示を変える処理の書き方を修正**
1e2326c19346941a9f1f592f193c0bb37774efa3

summertree128 commented on src/List.js:
```diff
@@ -0,0 +1,54 @@
+    updateIsEditing(true);
+  }
+
+  function handleMemoDetailLinkClick(e, m) {
```
`e`はReactで一般的に使われているので良いとして、`m`はぱっと見なんなのか分かりませんね。変に略さない方が良いと思います。

GennyBoy replied:
ありがとうございます！修正しました！
1b6a9393ef44bb5a48c955eb30cc1d9e03cd2442

summertree128 reviewed (commented):
修正ありがとうございます。いくつか追加でコメントしましたのでご確認ください〜

GennyBoy committed:
一般的な書き方に修正

GennyBoy committed:
ある条件を満たすときだけ要素を表示する際の通例の書き方に変更

GennyBoy committed:
関数の引数の名前をぱっと見で分かるように修正

GennyBoy committed:
必要なデータだけ localStorage から取得できるデータ設計に変更

やったこと
- localStorage に保存する際に、 key を `react-memo-spa` (このアプリ名) 、 value をメモのリストのオブジェクトを stringify したものにした
- 上記に合わせて、データの保存処理周りを変更

summertree128 commented on src/App.js:
```diff
@@ -10,29 +10,18 @@ export default function App() {
-        });
-      }
-    }
+    const memoList = JSON.parse(localStorage.getItem("react-memo-spa"));
```
以前のデータ設計よりはこちらの方が断然良いと思います〜 👍 
ただキーの名前にspaとかreactとかつくのはなんか変な気がするので、`memo`とかで良さそうです。
(Local Storageはドメインごとに独立して存在するものなので、アプリ名をキーで指定する必要はないです）

summertree128 reviewed (approved):
ご対応ありがとうございます！良い感じなったと思います。OKです〜
